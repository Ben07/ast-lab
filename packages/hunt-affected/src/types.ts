import { ParserOptions } from '@babel/parser';

export type Options = {
  /** Customised resolver function to resolve a module import path to absolute file path. */
  resolver?: (base: string, target: string) => Promise<string | void>;
  /** Function for reading file content from absolute file path. Uses fs.readFileSync to read with decoding *utf8* by default. */
  loader?: (path: string) => Promise<string | void>;
  /** Options provided by `@babel/parser`. Allow customize babel parser options while parsing file content to AST. */
  parserOptions?: ParserOptions;
};
