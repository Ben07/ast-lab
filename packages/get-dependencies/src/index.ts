import getDependencies from './getDependencies';

export { default as fileDepMap } from './fileDepMap';
export { default as mergeDepMap } from './mergeDepMap';
export { default as visitDepMap } from './visitDepMap';
export default getDependencies;