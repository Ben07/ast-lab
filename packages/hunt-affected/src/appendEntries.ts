import { DependencyMap, Module, Entry, AffectedMap, Member } from "ast-lab-types";

export default function appendEntries(depMap: DependencyMap, mod: Module, member: Member, entry: Entry[]): void {
  const affected  = depMap.get(mod) || new Map() as AffectedMap;
  if (!depMap.has(mod)) {
    depMap.set(mod, affected);
  }
  affected.set(member, affected.get(member) ? (affected.get(member) as Entry[]).concat(entry) : entry);
}