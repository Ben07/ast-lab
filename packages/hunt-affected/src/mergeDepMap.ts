import fs from 'fs';
import path from 'path';
import enhancedResolve from 'enhanced-resolve';
import _debug from 'debug';
import ecmaStats from 'es-stats';
import denodeify from './denodeify';
import { Import } from 'ast-lab-types';
import { Options, DeclarationNode, Resolver, DependencyMap } from './types';
import completeExtensions from './completeExtensions';
import resolveModule from './resolveModule';

const debug = _debug('hunt-affected:merge');

function defaultLoader(filePath: string): Promise<string> {
  return Promise.resolve(fs.readFileSync(filePath, 'utf8'));
}

async function convertImports(
  resolver: Resolver,
  baseDir: string,
  imports: Import[]
): Promise<{
  [key: string]: Import;
}> {
  const resolvedImports = await Promise.all(
    imports.map(async importInfo => {
      const { alias, source } = importInfo;
      const resolvedSource = await resolveModule(resolver, source, baseDir);
      return [alias, { ...importInfo, source: resolvedSource }];
    })
  );
  return Object.fromEntries(resolvedImports);
}

function getDeclaration(
  depMap: DependencyMap,
  source: string,
  declaration: string
): DeclarationNode {
  if (!depMap[source]) depMap[source] = {};
  if (!depMap[source][declaration]) {
    return (depMap[source][declaration] = {
      source,
      name: declaration,
      affects: [],
    });
  }
  return depMap[source][declaration];
}

type Extends = { [source: string]: Set<string> };
function handleExtends(depMap: DependencyMap, extendsToHandle: Extends): void {
  const remaining = { ...extendsToHandle };
  let current: any;
  let queue = Object.keys(extendsToHandle).reduce((ret, source) => {
    extendsToHandle[source].forEach(extend =>
      ret.push({
        source,
        extend,
      })
    );
    return ret;
  }, [] as Array<{ source: string; extend: string }>);
  let queueLen = queue.length;
  let nextBatch = [] as Array<{ source: string; extend: string }>;
  while ((current = queue.shift())) {
    const { source, extend } = current;
    if (depMap[extend]) {
      /** @todo check the extend priority logic */
      depMap[source] = {
        ...depMap[extend],
        ...depMap[source],
      };
    }
    if (remaining[extend]) {
      nextBatch.push(current);
    } else {
      remaining[source].delete(extend);
      if (!remaining[source].size) {
        delete remaining[source];
      }
    }
    if (!queue.length && queueLen > nextBatch.length) {
      queue = nextBatch;
      queueLen = nextBatch.length;
      nextBatch = [];
    }
  }
}

/**
 * Extract the dependency map of a list of source files.
 * @param sources A list of absolute source file paths
 * @param opts
 */
export default async function mergeDepMap(
  sources: string[],
  opts: Options = {}
): Promise<DependencyMap> {
  debug('===');
  const depMap = {} as DependencyMap;
  const parserOptions = opts?.parserOptions;
  const resolver =
    opts.resolver ||
    denodeify(
      enhancedResolve.create({
        ...(opts.resolverOptions || {}),
        extensions:
          opts.resolverOptions?.extensions ||
          completeExtensions(opts.parserOptions?.plugins),
      })
    );
  const loader = opts.loader || defaultLoader;

  const extendsToHandle = {} as Extends;
  const fileHandlers = sources.map(async src => {
    const fileContent = await loader(src);
    let fileStats;
    try {
      if (fileContent) {
        fileStats = ecmaStats(fileContent, parserOptions);
      }
    } catch (e) {
      console.warn(
        `get es stats from ${src} failed! Parser options: ${JSON.stringify(
          parserOptions
        )}`
      );
      console.warn(e);
    }
    if (!fileStats) return;
    const { relations, exports } = fileStats;
    const baseDir = path.dirname(src);
    const imports = convertImports(resolver, baseDir, fileStats.imports);
    if (exports.extends) {
      const modulePaths = await Promise.all(
        exports.extends.map(ex => resolveModule(resolver, ex, baseDir))
      );
      extendsToHandle[src] = new Set(modulePaths);
    }

    const declarations = Object.keys(relations);
    for (const declaration of declarations) {
      const declarationNode = getDeclaration(depMap, src, declaration);
      for (const dep of relations[declaration]) {
        if (typeof dep === 'string') {
          const isImported = imports[dep];
          if (isImported) {
            const { name, source } = isImported;
            getDeclaration(depMap, source, name).affects.push(declarationNode);
          } else {
            getDeclaration(depMap, src, dep).affects.push(declarationNode);
          }
        } else {
          const { source, name } = dep;
          const resolvedSource = await resolveModule(resolver, source, baseDir);
          getDeclaration(depMap, resolvedSource, name).affects.push(
            declarationNode
          );
        }
      }
    }
  });
  await Promise.all(fileHandlers);
  handleExtends(depMap, extendsToHandle);

  return depMap;
}
