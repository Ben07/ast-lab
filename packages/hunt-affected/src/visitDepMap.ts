import { DependencyMap, Entry, Member } from "ast-lab-types";
import _debug from 'debug';
import { MODULE_ALL } from "./constants";

const debug = _debug('hunt-affected:visit');

export type Visited = { [module: string]: Set<Member> };

/**
 * Walk through the dependency map from given entries to find out what are affected.
 * @param dependencyMap 
 * @param entries 
 */
export default function visitDependencyMap(dependencyMap: DependencyMap, entries: Entry[]): Visited {
  const visited = {} as Visited;
  let entryQueue = entries;
  debug('visit entries:', entries);
  while(entryQueue.length) {
    const { source: mod, name } = entryQueue.shift() as Entry;
    let modVisited = visited[mod];
    if (!modVisited) {
      modVisited = visited[mod] = new Set();
    }
    const affected = dependencyMap.get(mod);
    debug(`${mod} affects`, affected);
    if (affected && !modVisited.has(name)) {
      modVisited.add(name);
      const matchedEntries = affected.get(name);
      debug('current queue:', entryQueue);
      if (matchedEntries && matchedEntries.length) {
        debug('new to queue:', matchedEntries);
        entryQueue = entryQueue.concat(matchedEntries);
      }
      const allEntries = affected.get(MODULE_ALL);
      if (allEntries &&  allEntries.length) {
        debug('new to queue:', allEntries);
        entryQueue = entryQueue.concat(allEntries);
      }
    }
  }
  return visited;
}