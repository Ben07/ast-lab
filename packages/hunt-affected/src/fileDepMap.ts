import fs from 'fs';
import ecmaStats from 'es-stats';
import { Entry, Import, DependencyMap } from 'ast-lab-types';
import _debug from 'debug';
import appendEntries from './appendEntries';
import { MODULE_ALL } from './constants';
import { Options } from './types';

const debug = _debug('hunt-affected:file');
const dynamicImportReg = /[^#]+#/;

/**
 * Get a file's dependency map
 * @param filePath Absolute file path
 * @param options
 */
export default async function fileDepMap(filePath: string, { loader, parserOptions }: Options = {}): Promise<DependencyMap> {
  const depMap = new Map() as DependencyMap;
  const _loader = loader || ((_filePath: string) => Promise.resolve(fs.readFileSync(_filePath, 'utf8')));
  let file;
  try {
    file = await _loader(filePath);
    if (!file) {
      console.warn(`${filePath} is empty!`);
      return depMap;
    }
  } catch(e) {
    console.warn(`Failed to load file ${filePath}!`);
    return depMap;
  }
  let fileStats;
  try { 
    fileStats = ecmaStats(file, parserOptions);
  } catch(e) {
    console.warn(`get es stats from ${filePath} failed! Parser options: ${JSON.stringify(parserOptions)}`);
    console.warn(e);
    return depMap;
  }
  const { imports: target, exports: entry, relations } = fileStats;
  debug('>>> ', filePath);
  debug('imports:', target);
  debug('exports:', entry);
  debug('relations:', relations);
  
  const dynamicImportRenameQueue = [] as Array<Import>;
  const targetIndex = {} as { [key: string]: Import };
  target.forEach((ref) => {
    const { name, source, alias } = ref;
    targetIndex[alias] = ref;
    appendEntries(depMap, source, name, [] as Entry[]);
    if (dynamicImportReg.test(name)) {
      dynamicImportRenameQueue.push(ref);
    }
  });

  if (entry.extends) {
    entry.extends.forEach((source) => {
      appendEntries(depMap, source, MODULE_ALL, [{
        name: MODULE_ALL,
        source: filePath
      }]);
    })
  }

  debug('targetIndex:', targetIndex);
  debug('depMap initialized:', depMap);

  entry.members.forEach(({ alias }) => {
    const dependents = new Set() as Set<Import>;
    let dependsOn = relations[alias];
    if (dependsOn) {
      dependsOn.push(alias);
    } else {
      dependsOn = [alias];
    }
    const toClear = new Set() as Set<string>;
    dependsOn.forEach(dependent => {
      const importRef = targetIndex[dependent];
      if (importRef) {
        dependents.add(importRef);
      } else if (relations[dependent]) {
        dependsOn = dependsOn.concat(relations[dependent]);
        toClear.add(dependent);
      }
    });
    /** @todo verify whether this helps for performance */
    relations[alias] = dependsOn.filter(name => !toClear.has(name));

    dependents.forEach(importRef => {
      debug('importRef: ', importRef);
      appendEntries(depMap, importRef.source, importRef.name, [{
        name: alias,
        source: filePath
      }]);
    });

    // Dynamic import has special format name to avoid conflict
    // But this is not needed in file dep map
    dynamicImportRenameQueue.forEach(({ source, name }) => {
      const affectedMap = depMap.get(source);
      const entries = affectedMap && affectedMap.get(name);
      if (!affectedMap || !entries) { return; }
      const renameTo = name.replace(dynamicImportReg, '');
      appendEntries(depMap, source, renameTo, entries);
      affectedMap.delete(name);
    });

    // Add its own exports to the depMap
    appendEntries(depMap, filePath, alias, []);
  });
  debug('depMap', depMap);

  return depMap;
}
