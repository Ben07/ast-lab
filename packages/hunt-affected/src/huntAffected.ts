import { Options, Entry } from 'ast-lab-types';
import _debug from 'debug';
import mergeDepMap from './mergeDepMap';
import { default as visitDependencyMap, Visited as Affected } from './visitDepMap';

const debug = _debug('hunt-affected:affected');

export default async function huntAffected(sources: string[], entries: Entry[], opts: Options = {}): Promise<Affected> {
  debug('entries:', entries);
  const depMap = await mergeDepMap(sources, opts);
  debug('depMap:', depMap);
  return visitDependencyMap(depMap, entries);
}