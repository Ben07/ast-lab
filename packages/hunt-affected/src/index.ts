import huntAffected from './huntAffected';

export { Affected } from './visitDepMap';
export default huntAffected;
