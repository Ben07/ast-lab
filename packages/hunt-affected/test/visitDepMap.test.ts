import visitDepMap from '../src/visitDepMap';
import { DependencyMap } from '../src/types';

describe('visitDepMap()', () => {
  test('import default simple', () => {
    const c = {
      default: {
        source: '/c',
        name: 'default',
        affects: [],
      },
    };
    const b = {
      default: {
        source: '/b',
        name: 'default',
        affects: [c.default],
      },
    };
    const a = {
      default: {
        source: '/a',
        name: 'default',
        affects: [b.default],
      },
    };
    const depMap: DependencyMap = {
      '/a': a,
      '/b': b,
      '/c': c,
    };
    expect(
      visitDepMap(depMap, [{ name: 'default', source: '/a' }])
    ).toMatchSnapshot();
  });

  test('import multiples', () => {
    const depMap: DependencyMap = {
      '/a': {
        default: {
          source: '/a',
          name: 'default',
          affects: [],
        },
      },
      '/b': {
        default: {
          source: '/b',
          name: 'default',
          affects: [],
        },
      },
      '/c': {
        default: {
          source: '/c',
          name: 'default',
          affects: [],
        },
        c1: {
          source: '/c',
          name: 'c1',
          affects: [],
        },
        c2: {
          source: '/c',
          name: 'c2',
          affects: [],
        },
      },
      '/d': {
        default: {
          source: '/d',
          name: 'default',
          affects: [],
        },
      },
      '/e': {
        default: {
          source: '/e',
          name: 'default',
          affects: [],
        },
      },
    };
    depMap['/a'].default.affects.push(depMap['/b'].default);
    depMap['/b'].default.affects.push(depMap['/c'].c1);
    depMap['/b'].default.affects.push(depMap['/c'].c2);
    depMap['/c'].c1.affects.push(depMap['/d'].default);
    depMap['/c'].c1.affects.push(depMap['/e'].default);
    expect(
      visitDepMap(depMap, [
        {
          name: 'default',
          source: '/a',
        },
      ])
    ).toMatchSnapshot();
  });

  test('circular', () => {
    const depMap: DependencyMap = {
      '/a': {
        default: {
          source: '/a',
          name: 'default',
          affects: [],
        },
      },
      '/b': {
        default: {
          source: '/b',
          name: 'default',
          affects: [],
        },
      },
      '/c': {
        default: {
          source: '/c',
          name: 'default',
          affects: [],
        },
      },
    };
    depMap['/a'].default.affects.push(depMap['/b'].default);
    depMap['/b'].default.affects.push(depMap['/c'].default);
    depMap['/c'].default.affects.push(depMap['/a'].default);
    expect(
      visitDepMap(depMap, [{ name: 'default', source: '/a' }])
    ).toMatchSnapshot();
  });

  test('import all', () => {
    const depMap: DependencyMap = {
      '/a': {
        '*': {
          source: '/a',
          name: '*',
          affects: [],
        },
        default: {
          source: '/a',
          name: 'default',
          affects: [],
        },
      },
      '/b': {
        default: {
          source: '/b',
          name: 'default',
          affects: [],
        },
      },
      '/c': {
        default: {
          source: '/c',
          name: 'default',
          affects: [],
        },
      },
    };
    depMap['/a']['*'].affects.push(depMap['/b'].default);
    depMap['/b'].default.affects.push(depMap['/c'].default);
    expect(
      visitDepMap(depMap, [{ name: 'default', source: '/a' }])
    ).toMatchSnapshot();
  });
});
