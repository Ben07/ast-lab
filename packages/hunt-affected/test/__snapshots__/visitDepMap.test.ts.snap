// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`visitDepMap() circular 1`] = `
Object {
  "/a": Set {
    "default",
  },
  "/b": Set {
    "default",
  },
  "/c": Set {
    "default",
  },
}
`;

exports[`visitDepMap() import all 1`] = `
Object {
  "/a": Set {
    "default",
  },
  "/b": Set {
    "default",
  },
  "/c": Set {
    "default",
  },
}
`;

exports[`visitDepMap() import default simple 1`] = `
Object {
  "/a": Set {
    "default",
  },
  "/b": Set {
    "default",
  },
  "/c": Set {
    "default",
  },
}
`;

exports[`visitDepMap() import multiples 1`] = `
Object {
  "/a": Set {
    "default",
  },
  "/b": Set {
    "default",
  },
  "/c": Set {
    "c1",
    "c2",
  },
  "/d": Set {
    "default",
  },
  "/e": Set {
    "default",
  },
}
`;
